src/platform/Darwin/DnssdContexts.cpp

```diff
@@ -0,0 +1,426 @@
+}
+
+void RegisterContext::DispatchSuccess()
+{
+    callback(context, mType, CHIP_NO_ERROR);
+}
+
+BrowseContext::BrowseContext(void * cbContext, DnssdBrowseCallback cb, DnssdServiceProtocol cbContextProtocol)
+{
+    type     = ContextType::Browse;
+    context  = cbContext;
+    callback = cb;
+    protocol = cbContextProtocol;
+}
+
+void BrowseContext::DispatchFailure(DNSServiceErrorType err)
+{
+    ChipLogError(DeviceLayer, "Browse (%s)", Error::ToString(err));
+    callback(context, nullptr, 0, CHIP_ERROR_INTERNAL);
+    MdnsContexts::GetInstance().Remove(this);
+}
+
+void BrowseContext::DispatchSuccess()
+{
+    callback(context, services.data(), services.size(), CHIP_NO_ERROR);
+    MdnsContexts::GetInstance().Remove(this);
+}
+
+ResolveContext::ResolveContext(void * cbContext, DnssdResolveCallback cb, chip::Inet::IPAddressType cbAddressType)
+{
+    type     = ContextType::Resolve;
+    context  = cbContext;
+    callback = cb;
+    protocol = GetProtocol(cbAddressType);
+}
+
+ResolveContext::~ResolveContext()
+{
+    RemoveInterfaces();
+}
+
+void ResolveContext::DispatchFailure(DNSServiceErrorType err)
+{
+    ChipLogError(DeviceLayer, "Resolve (%s)", Error::ToString(err));
+    callback(context, nullptr, Span<Inet::IPAddress>(), CHIP_ERROR_INTERNAL);
+    MdnsContexts::GetInstance().Remove(this);
+}
+
+void ResolveContext::DispatchSuccess()
+{
+    for (auto & service : services)
+    {
+        auto & ips = addresses[service.first];
+
+        // Some interface may not have any ips, just ignore them.
+        if (ips.size() == 0)
+        {
+            continue;
+        }
+
+        // Use the first IP we got for the DnssdService.
+        service.second.mAddress.SetValue(ips.front());
+        ips.erase(ips.begin());
+        callback(context, &service.second, Span<Inet::IPAddress>(ips.data(), ips.size()), CHIP_NO_ERROR);
+        MdnsContexts::GetInstance().Remove(this);
+        return;
+    }
+}
+
+CHIP_ERROR ResolveContext::OnNewAddress(uint32_t interfaceId, const struct sockaddr * address)
+{
+    chip::Inet::IPAddress ip;
+    ReturnErrorOnFailure(chip::Inet::IPAddress::GetIPAddressFromSockAddr(*address, ip));
+    addresses[interfaceId].push_back(ip);
+
+#ifdef CHIP_DETAIL_LOGGING
+    char addrStr[INET6_ADDRSTRLEN];
+    ip.ToString(addrStr, sizeof(addrStr));
+    ChipLogDetail(DeviceLayer, "Mdns: %s interface: %" PRIu32 " ip:%s", __func__, interfaceId, addrStr);
+#endif // CHIP_DETAIL_LOGGING
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ResolveContext::OnNewLocalOnlyAddress()
+{
+    sockaddr_in6 sockaddr;
+    memset(&sockaddr, 0, sizeof(sockaddr));
+    sockaddr.sin6_len    = sizeof(sockaddr);
+    sockaddr.sin6_family = AF_INET6;
+    sockaddr.sin6_addr   = in6addr_loopback;
+    sockaddr.sin6_port   = htons((unsigned short) services[kDNSServiceInterfaceIndexLocalOnly].mPort);
+
+    return OnNewAddress(kDNSServiceInterfaceIndexLocalOnly, reinterpret_cast<struct sockaddr *>(&sockaddr));
```

Comment:
This used to use "any" for the interface with a comment about how we leave the routing to the network stack...  Is the change to local-only here purposeful?

Reply:
That was on purpose but maybe this is opening the risk to regress https://github.com/project-chip/connectedhomeip/pull/11992 
I have updated the code to use `0` just in case.